File: resources/ios/SecureStorageFunctions.swift
import Foundation
import Security

enum SecureStorageFunctions {

    private static let service = "com.sandip.plugins.secure_storage"

    private static let maxKeyLength = 255
    private static let maxValueLength = 8192

    private static func baseQuery(forKey key: String) -> [String: Any] {
        return [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
    }

    private static func validateKey(_ parameters: [String: Any]) -> Result<String, [String: Any]> {
        guard let rawKey = parameters["key"] as? String else {
            return .failure(BridgeResponse.error(code: "KEY_REQUIRED", message: "A non-empty key is required."))
        }

        let key = rawKey.trimmingCharacters(in: .whitespacesAndNewlines)

        if key.isEmpty {
            return .failure(BridgeResponse.error(code: "KEY_REQUIRED", message: "A non-empty key is required."))
        }

        if key.count > maxKeyLength {
            return .failure(BridgeResponse.error(code: "KEY_TOO_LONG", message: "Key must not exceed \(maxKeyLength) characters."))
        }

        return .success(key)
    }

    class Set: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            switch SecureStorageFunctions.validateKey(parameters) {
            case .failure(let error):
                return error
            case .success(let key):
                var query = SecureStorageFunctions.baseQuery(forKey: key)

                guard let value = parameters["value"] as? String else {
                    SecItemDelete(query as CFDictionary)
                    return BridgeResponse.success(data: ["success": true])
                }

                if value.utf8.count > SecureStorageFunctions.maxValueLength {
                    return BridgeResponse.error(code: "VALUE_TOO_LARGE", message: "Value must not exceed \(SecureStorageFunctions.maxValueLength) bytes.")
                }

                let data = Data(value.utf8)

                let attributesToUpdate: [String: Any] = [kSecValueData as String: data]
                let updateStatus = SecItemUpdate(query as CFDictionary, attributesToUpdate as CFDictionary)

                if updateStatus == errSecItemNotFound {
                    query[kSecValueData as String] = data
                    query[kSecAttrAccessible as String] = kSecAttrAccessibleWhenUnlockedThisDeviceOnly
                    let addStatus = SecItemAdd(query as CFDictionary, nil)

                    guard addStatus == errSecSuccess else {
                        return BridgeResponse.error(code: "WRITE_FAILED", message: "Could not write to Keychain (status \(addStatus)).")
                    }
                } else if updateStatus != errSecSuccess {
                    return BridgeResponse.error(code: "WRITE_FAILED", message: "Could not update Keychain item (status \(updateStatus)).")
                }

                return BridgeResponse.success(data: ["success": true])
            }
        }
    }

    class Get: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            switch SecureStorageFunctions.validateKey(parameters) {
            case .failure(let error):
                return error
            case .success(let key):
                var query = SecureStorageFunctions.baseQuery(forKey: key)
                query[kSecReturnData as String] = true
                query[kSecMatchLimit as String] = kSecMatchLimitOne

                var item: CFTypeRef?
                let status = SecItemCopyMatching(query as CFDictionary, &item)

                guard status == errSecSuccess, let data = item as? Data,
                      let value = String(data: data, encoding: .utf8) else {
                    return BridgeResponse.success(data: ["value": ""])
                }

                return BridgeResponse.success(data: ["value": value])
            }
        }
    }

    class Delete: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            switch SecureStorageFunctions.validateKey(parameters) {
            case .failure(let error):
                return error
            case .success(let key):
                let status = SecItemDelete(SecureStorageFunctions.baseQuery(forKey: key) as CFDictionary)

                guard status == errSecSuccess || status == errSecItemNotFound else {
                    return BridgeResponse.error(code: "DELETE_FAILED", message: "Could not delete Keychain item (status \(status)).")
                }

                return BridgeResponse.success(data: ["success": true])
            }
        }
    }
}

File: resources/android/SecureStorageFunctions.kt
package com.sandip.plugins.secure_storage

import android.content.Context
import android.content.SharedPreferences
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import java.security.GeneralSecurityException
import java.security.KeyStore
import java.security.MessageDigest
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

object SecureStorageFunctions {

    private const val PREFS_NAME = "sghimire_secure_storage_prefs"
    private const val ANDROID_KEYSTORE = "AndroidKeyStore"
    private const val KEY_ALIAS = "sghimire_secure_storage_master_key"
    private const val TRANSFORMATION = "AES/GCM/NoPadding"
    private const val GCM_TAG_LENGTH_BITS = 128
    private const val GCM_IV_LENGTH_BYTES = 12

    private const val MAX_KEY_LENGTH = 255
    private const val MAX_VALUE_LENGTH = 8192

    private fun prefs(context: Context): SharedPreferences =
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private fun secretKey(): SecretKey {
        val keyStore = KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }

        (keyStore.getKey(KEY_ALIAS, null) as? SecretKey)?.let {
            return it
        }

        val keyGenerator =
                KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE)
        keyGenerator.init(
                KeyGenParameterSpec.Builder(
                                KEY_ALIAS,
                                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
                        )
                        .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                        .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                        .setKeySize(256)
                        .build()
        )
        return keyGenerator.generateKey()
    }

    private fun storageKey(key: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(key.toByteArray(Charsets.UTF_8))
        return Base64.encodeToString(digest, Base64.NO_WRAP or Base64.URL_SAFE)
    }

    private fun encrypt(plainText: String): String {
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, secretKey())
        val cipherText = cipher.doFinal(plainText.toByteArray(Charsets.UTF_8))
        return Base64.encodeToString(cipher.iv + cipherText, Base64.NO_WRAP)
    }

    private fun decrypt(encoded: String): String {
        val combined = Base64.decode(encoded, Base64.NO_WRAP)
        val iv = combined.copyOfRange(0, GCM_IV_LENGTH_BYTES)
        val cipherText = combined.copyOfRange(GCM_IV_LENGTH_BYTES, combined.size)

        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, secretKey(), GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        return String(cipher.doFinal(cipherText), Charsets.UTF_8)
    }

    private fun validateKey(parameters: Map<String, Any>): Pair<String?, Map<String, Any>?> {
        val key = (parameters["key"] as? String)?.trim()

        if (key.isNullOrEmpty()) {
            return null to BridgeResponse.error("KEY_REQUIRED", "A non-empty key is required.")
        }

        if (key.length > MAX_KEY_LENGTH) {
            return null to
                    BridgeResponse.error(
                            "KEY_TOO_LONG",
                            "Key must not exceed $MAX_KEY_LENGTH characters."
                    )
        }

        return key to null
    }

    class Set(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val (key, keyError) = validateKey(parameters)
            if (keyError != null) return keyError

            val value = parameters["value"] as? String

            if (value != null && value.toByteArray(Charsets.UTF_8).size > MAX_VALUE_LENGTH) {
                return BridgeResponse.error(
                        "VALUE_TOO_LARGE",
                        "Value must not exceed $MAX_VALUE_LENGTH bytes."
                )
            }

            return try {
                val editor = prefs(activity).edit()
                if (value == null) {
                    editor.remove(storageKey(key!!))
                } else {
                    editor.putString(storageKey(key!!), encrypt(value))
                }

                if (editor.commit()) {
                    BridgeResponse.success(mapOf("success" to true))
                } else {
                    BridgeResponse.error("WRITE_FAILED", "Could not write to secure storage.")
                }
            } catch (e: GeneralSecurityException) {
                BridgeResponse.error(
                        "WRITE_FAILED",
                        e.message ?: "Could not access the secure keystore."
                )
            } catch (e: Exception) {
                BridgeResponse.error(
                        "WRITE_FAILED",
                        e.message ?: "Could not write to secure storage."
                )
            }
        }
    }

    class Get(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val (key, keyError) = validateKey(parameters)
            if (keyError != null) return keyError

            return try {
                val stored = prefs(activity).getString(storageKey(key!!), null)
                BridgeResponse.success(mapOf("value" to (stored?.let { decrypt(it) } ?: "")))
            } catch (e: GeneralSecurityException) {
                BridgeResponse.error(
                        "READ_FAILED",
                        e.message ?: "Could not access the secure keystore."
                )
            } catch (e: Exception) {
                BridgeResponse.error(
                        "READ_FAILED",
                        e.message ?: "Could not read from secure storage."
                )
            }
        }
    }

    class Delete(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val (key, keyError) = validateKey(parameters)
            if (keyError != null) return keyError

            return try {
                if (prefs(activity).edit().remove(storageKey(key!!)).commit()) {
                    BridgeResponse.success(mapOf("success" to true))
                } else {
                    BridgeResponse.error("DELETE_FAILED", "Could not delete from secure storage.")
                }
            } catch (e: GeneralSecurityException) {
                BridgeResponse.error(
                        "DELETE_FAILED",
                        e.message ?: "Could not access the secure keystore."
                )
            } catch (e: Exception) {
                BridgeResponse.error(
                        "DELETE_FAILED",
                        e.message ?: "Could not delete from secure storage."
                )
            }
        }
    }
}
